### ISSUE TITLE
Mask "shadows" in some images?

### ISSUE THREAD

**Issue by eduardathome** [saic-mdal/lama#78]
In some predicted images, there is a noticeable marking of the applied masks. I attached a small example here. This is part of a bigger image. https://imgur.com/a/19pIbQo 
I circled in red the "shadows" that I'm referring to. Those are exactly where the random masks were applied. 

Worth mentioning: This is a model that was trained on my own data, using the proposed architecture and configuration proposed here: https://github.com/saic-mdal/lama/blob/main/configs/training/lama-regular.yaml

I am not certain what other information is relevant here but I will provide more if necessary. 

Any suggestions on what the issue is here and what could fix it are welcomed.  Thanks and thank you for the great project.

**Comment by windj007** [comment]
Hi! Sorry for the late reply.

Does this kind of artifacts appear always? Or maybe there are specific conditions where they are more noticeable than usual?

Basically, you can try setting `training_model.image_to_discriminator=inpainted` - thus the discriminator and the losses will see a blended image `gen_output * mask + input * (1 - mask)` instead of raw `gen_output`.

**Comment by windj007** [comment]
How big is your dataset? What is your training resolution and testing resolution?

**Comment by eduardathome** [comment]
Hi, Sorry for the even later reply. 

> Does this kind of artifacts appear always? Or maybe there are specific conditions where they are more noticeable than usual?

It doesn't always appear: it is more present when the mask is over a solid color than on textures and meaningful content. 

> How big is your dataset? What is your training resolution and testing resolution? 

The training set has 150000 images. The validation and test sets are 300 each. Resolution vary, with largest side being 1500px. 

> Basically, you can try setting training_model.image_to_discriminator=inpainted

I did one experiment so far, resuming the training on a trained model (for 100 epochs) and after 3, 10 and 30 epochs, the artifacts did not improve. 

**_I also tested the proposed model big-lama and I encountered the same artifacts, using the proposed bin/predict.py script. I will attach 3 conclusive results. If you want to replicate this, let me know how you'd like me to attach the inputs._** 
![1](https://user-images.githubusercontent.com/26660549/152366486-c62e57ae-7a54-4252-83e1-05fe39aa2802.jpeg)
![8](https://user-images.githubusercontent.com/26660549/152366490-4f1afa76-07a6-491f-94ee-544655fc8977.jpeg)
![11](https://user-images.githubusercontent.com/26660549/152366491-413406d5-92c5-45d3-a9f1-cac9194d82a0.jpeg)

**Comment by windj007** [comment]
Hm... that's weird

In which resolution do you feed the images during training? Am I correct that you're applying Lama-Regular model to images of much higher resolution than during training? 

> I also tested the proposed model big-lama

Do you mean that a pre-trained (by us on Places) Big-Lama has same artifacts? Or you re-trained it on your data?

> If you want to replicate this, let me know how you'd like me to attach the inputs.

Yes, a dozen of relevant images+masks would help a lot!

Thank you!

**Comment by eduardathome** [comment]
Hi again,

> In which resolution do you feed the images during training? Am I correct that you're applying Lama-Regular model to images of much higher resolution than during training?

You are right, I resize images to 512x512 during training and yes, I use Lama-Regular configuration. During prediction I don't resize, using original size ( which varies but it is around 1500).

Your assumption is probably correct, since I tested the same prediction on 512x512 images and the artifacts, while not entirely gone, are significantly reduced. 

> Do you mean that a pre-trained (by us on Places) Big-Lama has same artifacts? Or you re-trained it on your data?

Yes, I get similar artifacts using the model proposed by you, more exactly: "The best model (Places2, Places Challenge):", downloaded from: 
```
curl -L $(yadisk-direct https://disk.yandex.ru/d/ouP6l8VJ0HpMZg) -o big-lama.zip
```
Below I attached 12 images and mask that should replicate the artifacts. If you prefer them in a different format, let me know. 

[artifacts.zip](https://github.com/saic-mdal/lama/files/8016390/artifacts.zip)

I will also follow up with 1-2 more general questions, so if you could, don't close the Issue yet. 

Thanks for taking the time.

**Comment by windj007** [comment]
Thank you! We'll check your images!

**Comment by ImmortalSdm** [comment]
@windj007  I'm facing with the same problem after training 40 epoch on FFHQ dataset. However,the  artifacts  are much more obvious, any idea to alleviate the problem?
https://s2.loli.net/2022/03/08/ZiBMduyKPJporVF.jpg

**Comment by windj007** [comment]
@ImmortalSdm is this image from training dataloader or from validation? If this is validation, then such artifacts may apper when the mask is non-binary, e.g. it has smooth transition between 0 (known areas) and 1 (missing areas).

**Comment by windj007** [comment]
@eduardathome Sorry for the late reply again!

Have you tried training in 256?

Already after releasing the codebase we found that with Fourier-based generators training in 256 yields more robust performance than training in higher resolutions - that's most probably due to characteristics of the loss functions.

**Comment by ImmortalSdm** [comment]
> @ImmortalSdm is this image from training dataloader or from validation? If this is validation, then such artifacts may apper when the mask is non-binary, e.g. it has smooth transition between 0 (known areas) and 1 (missing areas).



> @ImmortalSdm is this image from training dataloader or from validation? If this is validation, then such artifacts may apper when the mask is non-binary, e.g. it has smooth transition between 0 (known areas) and 1 (missing areas).

yep, it's from validation. Thanks for your reply, i will check my mask image.

**Comment by Marcelo5444** [comment]
Hi! I am facing some issues related with this. I am just fine tunning with LAMA. At first in order to do sannity check, I tried to just overfit to a single image of CelebHQ. When using the predict.py file, Everything works fine, but when training (overfitting) Saving the output of the network at different iterations I obtain this
In Here you can see the following. Top left - original image, Top right image with mask on it. bottom left output imagen of the network. Bottom right inpainted image
![epoch=1_0_legend](https://user-images.githubusercontent.com/44267105/160933469-f5eeb6e3-7992-4f0a-ba61-9521b3e8278b.png)
.

**Comment by windj007** [comment]
When fine-tuning a pretrained model, please keep in mind:
1. Overfitting to a single image with discriminator might break because it is too easy for discriminator to remember the real image exactly - so it wins and the training diverges
2. When we tried resuming training from a checkpoint, we encountered instabilities when batch size or number of gpus was changed after restart compared to those before. I do not know exactly why, but most probably due to batchnorm or adam stats.

**Comment by Agrechka** [comment]
@ImmortalSdm Hello! Did you manage to train lama on ffhq in the end? I am having issues and I would so greatly appreciate a trianed checkpoint if you managed.

**Comment by yftongbupt** [comment]
> Hi again,
> 
> > In which resolution do you feed the images during training? Am I correct that you're applying Lama-Regular model to images of much higher resolution than during training?
> 
> You are right, I resize images to 512x512 during training and yes, I use Lama-Regular configuration. During prediction I don't resize, using original size ( which varies but it is around 1500).
> 
> Your assumption is probably correct, since I tested the same prediction on 512x512 images and the artifacts, while not entirely gone, are significantly reduced.
> 
> > Do you mean that a pre-trained (by us on Places) Big-Lama has same artifacts? Or you re-trained it on your data?
> 
> Yes, I get similar artifacts using the model proposed by you, more exactly: "The best model (Places2, Places Challenge):", downloaded from:
> 
> ```
> curl -L $(yadisk-direct https://disk.yandex.ru/d/ouP6l8VJ0HpMZg) -o big-lama.zip
> ```
> 
> Below I attached 12 images and mask that should replicate the artifacts. If you prefer them in a different format, let me know.
> 
> [artifacts.zip](https://github.com/saic-mdal/lama/files/8016390/artifacts.zip)
> 
> I will also follow up with 1-2 more general questions, so if possible, don't close the Issue yet.
> 
> Thanks for taking the time.

@eduardathome Have you solved the problem? I have the same issues as you. I would so greatly appreciate if you share your solution.
